packages/scrapers-v2/src/lexer/lexer.ts

```diff
@@ -0,0 +1,432 @@
+    .map((link) => {
+      const url = new URL(link, base);
+      url.hash = concentrationsTag;
+      return url.toString();
+    });
+};
+
+/**
+ * Converts tables rows into a list of HRows
+ * @param $
+ * @param table
+ */
+const tokenizeRows = ($: CheerioStatic, table: CheerioElement): HRow[] => {
+  const courseTable: HRow[] = [];
+
+  for (const tr of $(table).find("tbody > tr").toArray()) {
+    // different row type
+    const type = getRowType($, tr);
+    const row = constructRow($, tr, type);
+    courseTable.push(row);
+  }
+
+  return courseTable;
+};
+
+/**
+ * Pre-parses the row to determine its type
+ * @param $
+ * @param tr
+ */
+const getRowType = ($: CheerioStatic, tr: CheerioElement) => {
+  const trClass = tr.attribs["class"];
+  const td = $(tr.children[0]);
+  const tdClass = $(tr).children()[0].attribs["class"];
+
+  if (trClass.includes("subheader")) {
+    return HRowType.SUBHEADER;
+  } else if (trClass.includes("areaheader")) {
+    return HRowType.HEADER;
+  } else if (trClass.includes("orclass")) {
+    return HRowType.OR_COURSE;
+  } else if (tdClass) {
+    if (tdClass !== "codecol") {
+      throw Error(
+        "Expected class to exist with value codecol. Please add this case to the compiler."
+      );
+    }
+    if (td.find(".code").toArray().length > 1) {
+      return HRowType.AND_COURSE;
+    }
+    return HRowType.PLAIN_COURSE;
+  }
+
+  const tdText = parseText(td);
+  // Different range types
+  if (
+    RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_1.test(tdText) ||
+    RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_2.test(tdText)
+  ) {
+    return HRowType.RANGE_LOWER_BOUNDED;
+  } else if (RANGE_BOUNDED.test(tdText)) {
+    return HRowType.RANGE_BOUNDED;
+  } else if (RANGE_UNBOUNDED.test(tdText)) {
+    return HRowType.RANGE_UNBOUNDED;
+  }
+
+  return HRowType.COMMENT;
+};
+
+/**
+ * Converts a single row based on the passed-in type (determined by {@link getRowType}
+ * @param $
+ * @param tr
+ * @param type
+ */
+const constructRow = (
+  $: CheerioStatic,
+  tr: CheerioElement,
+  type: HRowType
+): HRow => {
+  switch (type) {
+    case HRowType.HEADER:
+    case HRowType.SUBHEADER:
+    case HRowType.COMMENT:
+      return constructTextRow($, tr, type);
+    case HRowType.OR_COURSE:
+      return constructOrCourseRow($, tr);
+    case HRowType.PLAIN_COURSE:
+      return constructPlainCourseRow($, tr);
+    case HRowType.AND_COURSE:
+      return constructMultiCourseRow($, tr);
+    case HRowType.RANGE_LOWER_BOUNDED:
+      return constructRangeLowerBoundedMaybeExceptions($, tr);
+    case HRowType.RANGE_BOUNDED:
+      return constructRangeBounded($, tr);
+    case HRowType.RANGE_UNBOUNDED:
+      return constructRangeUnbounded($, tr);
+    case HRowType.RANGE_LOWER_BOUNDED_WITH_EXCEPTIONS:
+      throw "should never get here";
+    default:
+      return assertUnreachable(type);
+  }
+};
+
+const constructTextRow = <T>(
+  $: CheerioStatic,
+  tr: CheerioElement,
+  type: T
```

Comment:
Because for other functions `constructPlainCourseRow`, `constructOrCourseRow`, etc, their types are very specific, which would be just PLAINE_COURSE, OR, etc. But for `constructTextRow`, there are three different possibilities (in switch statement) which can be `HEADER`, `SUBHEADER`, or `COMMENT`, so we use generic type here

Reply:
ahhh i see ty!